x-pack/legacy/plugins/siem/server/lib/detection_engine/alerts/utils.test.ts

```diff
@@ -99,8 +103,45 @@ describe('utils', () => {
       });
     });
   });
-  describe('singleBulkIndex', () => {
-    test('create successful bulk index', async () => {
+  describe('singleBulkCreate', () => {
+    describe('create signal id gereateId', () => {
+      test('two docs with same index, id, and version should have same id', () => {
+        const findex = 'myfakeindex';
+        const fid = 'somefakeid';
+        const version = '1';
+        const firstHash = generateId(findex, fid, version);
+        const secondHash = generateId(findex, fid, version);
+        expect(firstHash).toEqual(secondHash);
+      });
+      test('two docs with different index, id, and version should have different id', () => {
+        const findex = 'myfakeindex';
+        const findex2 = 'mysecondfakeindex';
+        const fid = 'somefakeid';
+        const version = '1';
+        const firstHash = generateId(findex, fid, version);
+        const secondHash = generateId(findex2, fid, version);
+        expect(firstHash).not.toEqual(secondHash);
+      });
+      test('two docs with same index, different id, and same version should have different id', () => {
+        const findex = 'myfakeindex';
+        const fid = 'somefakeid';
+        const fid2 = 'somefakeid2';
+        const version = '1';
+        const firstHash = generateId(findex, fid, version);
+        const secondHash = generateId(findex, fid2, version);
+        expect(firstHash).not.toEqual(secondHash);
+      });
+      test('two docs with same index, same id, and different version should have different id', () => {
+        const findex = 'myfakeindex';
+        const fid = 'somefakeid';
+        const version = '1';
+        const version2 = '2';
+        const firstHash = generateId(findex, fid, version);
+        const secondHash = generateId(findex, fid, version2);
+        expect(firstHash).not.toEqual(secondHash);
+      });
+    });
+    test('create successful bulk create', async () => {
```

Comment:
Optional but suggestions for a few tests to help out:

1) Test the exact output:
const id = generateId(findex, fid, version);
expect(id).toEqual(...whatever the hex is...);

Reason being that you want to ensure that you have a stable non-changing ID and if anyone accidentally changes the output type (such as an upgrade to a newer NodeJS or a difference in the sha256 algorithm) your test will trigger that it is getting an unexpected hash. After all the hash should hash to the same thing always.

2) Test that the length is always below the limit for an ElasticSearch `_id`

The max length according to the docs:
https://www.elastic.co/guide/en/elasticsearch/reference/current/mapping-id-field.html

Looks to be 512 bytes which looks like it in the ES code base:
https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/action/index/IndexRequest.java#L200

I think you can do the same length trick that their code does with UTF-8 to check your hash lengths. Your test should have an incredibly long index like so for this test:

```ts
const id = generateId('index_that_is_so_long_it_will_really_mess_with_you_wow_how_long', fid, version);
expect(id.length) toBeLessThan(maxLength)
```

It looks like the length of your test index should be 255 for this test as that's the maximum length of an index:
https://discuss.elastic.co/t/index-name-type-name-and-field-name-rules/133039/2

If you do these two tests you should gain confidence that this works as you expect with worst case situations and if you need to change the hashing algorithm you won't have as many worries it will blow up later on.

Reply:
I agree, those are good variables to check for in the tests as well.